File: Services/DashboardService.cs
using controle_ja_mobile.Models;
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;

namespace controle_ja_mobile.Services
{
    public class DashboardService
    {
        private readonly ApiService _apiService;

        public DashboardService(ApiService apiService)
        {
            _apiService = apiService;
        }

        public async Task<List<ChartData>> getExpensesByCategory(long start, long end)
        {
            try
            {
                var response = await _apiService.GetAsync<string>($"dashboard/expenses-category?start={start}&end={end}");
                if (!string.IsNullOrWhiteSpace(response))
                {
                    var chartDataResponse = JsonSerializer.Deserialize<List<ChartData>>(response, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                    if (chartDataResponse != null)
                    {
                        return chartDataResponse;
                    }
                }
                return new List<ChartData>();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return new List<ChartData>();
            }
        }

        public async Task<List<ChartData>> getCreditExpensesByCategory(long start, long end)
        {
            try
            {
                var response = await _apiService.GetAsync<string>($"dashboard/credit-expenses-category?start={start}&end={end}");
                if (!string.IsNullOrWhiteSpace(response))
                {
                    var chartDataResponse = JsonSerializer.Deserialize<List<ChartData>>(response, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                    if (chartDataResponse != null)
                    {
                        return chartDataResponse;
                    }
                }
                return new List<ChartData>();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return new List<ChartData>();
            }
        }

        public async Task<List<ChartData>> getIncomesByCategory(long start, long end)
        {
            try
            {
                var response = await _apiService.GetAsync<string>($"dashboard/incomes-category?start={start}&end={end}");
                if (!string.IsNullOrWhiteSpace(response))
                {
                    var chartDataResponse = JsonSerializer.Deserialize<List<ChartData>>(response, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                    if (chartDataResponse != null)
                    {
                        return chartDataResponse;
                    }
                }
                return new List<ChartData>();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return new List<ChartData>();
            }
        }

        public async Task<List<ChartData>> getFuelComparison(long start, long end)
        {
            try
            {
                var response = await _apiService.GetAsync<string>($"dashboard/fuel-comparison?start={start}&end={end}");
                if (!string.IsNullOrWhiteSpace(response))
                {
                    var chartDataResponse = JsonSerializer.Deserialize<List<ChartData>>(response, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                    if (chartDataResponse != null)
                    {
                        return chartDataResponse;
                    }
                }
                return new List<ChartData>();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return new List<ChartData>();
            }
        }

        public async Task<FinancialSummary?> getDashboardSummary(long start, long end)
        {
            try
            {
                var response = await _apiService.GetAsync<string>($"dashboard/summary?start={start}&end={end}");
                if (!string.IsNullOrWhiteSpace(response))
                {
                    var financialSummary = JsonSerializer.Deserialize<FinancialSummary>(response, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                    if (financialSummary != null)
                    {
                        return financialSummary;
                    }
                }
                return new FinancialSummary();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return new FinancialSummary();
            }
        }

        public async Task<List<ChartData>> getEvolution(long start, long end, string uuid)
        {
            try
            {
                string endpoint = $"dashboard/evolution?start={start}&end={end}";
                if (!string.IsNullOrEmpty(uuid))
                {
                    endpoint += $"&categoryId={uuid}";
                }
                var response = await _apiService.GetAsync<string>(endpoint);

                if (!string.IsNullOrWhiteSpace(response))
                {
                    var chartDataResponse = JsonSerializer.Deserialize<List<ChartData>>(response, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                    if (chartDataResponse != null)
                    {
                        return chartDataResponse;
                    }
                }
                return new List<ChartData>();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return new List<ChartData>();
            }
        }

        public async Task<DashboardFullSummary?> GetFullSummaryAsync(long start, long end)
        {
            try
            {
                string endpoint = $"dashboard/full-summary?start={start}&end={end}";
                var response = await _apiService.GetAsync<string>(endpoint);

                if (!string.IsNullOrWhiteSpace(response))
                {
                    var options = new JsonSerializerOptions
                    {
                        PropertyNameCaseInsensitive = true,
                        NumberHandling = System.Text.Json.Serialization.JsonNumberHandling.AllowReadingFromString
                    };
                    var summary = JsonSerializer.Deserialize<DashboardFullSummary>(response, options);
                    if (summary != null) return summary;
                }
                return null;
            }
            catch (Exception ex)
            {
                Console.WriteLine("ERRO JSON DASHBOARD: " + ex.Message);
                return null;
            }
        }
    }
}

File: lib/services/dashboard_service.dart
import 'dart:convert';

import '../models/dashboard_data.dart';
import 'api_client.dart';

/// Serviço do dashboard.
///
/// Migrado de DashboardService.cs — mantém todas as rotas exatas da API.
class DashboardService {
  final ApiClient _apiClient;

  DashboardService({required ApiClient apiClient}) : _apiClient = apiClient;

  Future<List<ChartData>> getExpensesByCategory(int start, int end) async {
    try {
      final response = await _apiClient.dio
          .get('dashboard/expenses-category?start=$start&end=$end');
      return _parseChartDataList(response.data);
    } catch (e) {
      return [];
    }
  }

  Future<List<ChartData>> getCreditExpensesByCategory(
      int start, int end) async {
    try {
      final response = await _apiClient.dio
          .get('dashboard/credit-expenses-category?start=$start&end=$end');
      return _parseChartDataList(response.data);
    } catch (e) {
      return [];
    }
  }

  Future<List<ChartData>> getIncomesByCategory(int start, int end) async {
    try {
      final response = await _apiClient.dio
          .get('dashboard/incomes-category?start=$start&end=$end');
      return _parseChartDataList(response.data);
    } catch (e) {
      return [];
    }
  }

  Future<List<ChartData>> getFuelComparison(int start, int end) async {
    try {
      final response = await _apiClient.dio
          .get('dashboard/fuel-comparison?start=$start&end=$end');
      return _parseChartDataList(response.data);
    } catch (e) {
      return [];
    }
  }

  Future<FinancialSummary?> getDashboardSummary(int start, int end) async {
    try {
      final response = await _apiClient.dio
          .get('dashboard/summary?start=$start&end=$end');
      if (response.data != null) {
        final parsed = response.data is String
            ? jsonDecode(response.data as String) as Map<String, dynamic>
            : response.data as Map<String, dynamic>;
        return FinancialSummary.fromJson(parsed);
      }
      return FinancialSummary();
    } catch (e) {
      return FinancialSummary();
    }
  }

  Future<List<ChartData>> getEvolution(
      int start, int end, String? uuid) async {
    try {
      String endpoint = 'dashboard/evolution?start=$start&end=$end';
      if (uuid != null && uuid.isNotEmpty) {
        endpoint += '&categoryId=$uuid';
      }
      final response = await _apiClient.dio.get(endpoint);
      return _parseChartDataList(response.data);
    } catch (e) {
      return [];
    }
  }

  Future<DashboardFullSummary?> getFullSummary(
      int start, int end) async {
    try {
      final response = await _apiClient.dio
          .get('dashboard/full-summary?start=$start&end=$end');
      if (response.data != null) {
        final parsed = response.data is String
            ? jsonDecode(response.data as String) as Map<String, dynamic>
            : response.data as Map<String, dynamic>;
        return DashboardFullSummary.fromJson(parsed);
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  // --- Helper ---

  List<ChartData> _parseChartDataList(dynamic data) {
    if (data != null) {
      final List<dynamic> list = data is String
          ? jsonDecode(data) as List<dynamic>
          : data as List<dynamic>;
      return list
          .map((e) => ChartData.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return [];
  }
}
